File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/models/list/FCPListSection.kt
package com.oguzhnatly.flutter_carplay.models.list

import androidx.car.app.model.ItemList
import androidx.car.app.model.SectionedItemList

/**
 * A wrapper class for CPListSection with additional functionality.
 *
 * @param obj A map containing information about the list section.
 */
class FCPListSection
    (obj: Map<String, Any>) {

    /// The underlying CPListSection instance.
    private lateinit var _super: SectionedItemList

    /// The unique identifier for the list section.
    var elementId: String
        private set

    /// The header text for the list section (optional).
    private var header: String

    /// An array of CPListTemplateItem instances associated with the list section.
    private lateinit var items: ItemList

    /// An array of FCPListItem instances associated with the list section.
    private var objcItems: List<FCPListItem>

    init {
        val elementIdValue = obj["_elementId"] as? String?
        assert(elementIdValue != null) { "Missing required keys in dictionary for FCPListSection initialization." }
        elementId = elementIdValue!!
        header = obj["header"] as? String ?: ""
        objcItems = (obj["items"] as? List<Map<String, Any>> ?: emptyList<FCPListItem>()).map {
            FCPListItem(it as Map<String, Any>)
        }
        val builder = ItemList.Builder()
        objcItems.forEach {
            builder.addItem(it.getTemplate)
        }
        items = builder.build()
    }


    /** Returns the underlying CPListSection instance configured with the specified properties. */
    val getTemplate: SectionedItemList
        get() {
            val listSection = SectionedItemList.create(items, header)
            _super = listSection
            return listSection
        }


    /**
     * Retrieves an array of FCPListItem instances associated with the list section.
     *
     * @return An array of FCPListItem instances.
     */
    fun getItems(): List<FCPListItem> {
        return objcItems
    }
}

File: ios/flutter_carplay/Sources/flutter_carplay/models/list/FCPListSection.swift
//
//  FCPListSection.swift
//  flutter_carplay
//
//  Created by Oğuzhan Atalay on 21.08.2021.
//

import CarPlay

@available(iOS 14.0, *)
class FCPListSection {
  private(set) var _super: CPListSection?
  private(set) var elementId: String
  private var header: String?
  private var items: [CPListTemplateItem]
  private var objcItems: [FCPListTemplateItem]
  private var sectionIndexEnabled: Bool

  init(obj: [String: Any]) {
    self.elementId = obj["_elementId"] as! String
    self.header = obj["header"] as? String
    self.sectionIndexEnabled = obj["sectionIndexEnabled"] as? Bool ?? true
    self.objcItems = (obj["items"] as! [[String: Any]]).map { dict -> FCPListTemplateItem in
      guard let runtimeType = dict["runtimeType"] as? String else {
        fatalError("FCPListSection.init: Missing runtimeType in item")
      }

      if runtimeType == "FCPListImageRowItem" {
        return FCPListImageRowItem(obj: dict) as FCPListTemplateItem
      } else if runtimeType == "FCPListItem" {
        return FCPListItem(obj: dict) as FCPListTemplateItem
      } else {
        fatalError("FCPListSection.init: Unknown item runtimeType: \(runtimeType)")
      }
    }
    self.items = self.objcItems.map {
      $0.get
    }
  }

  var get: CPListSection {
    let sectionIndexTitle = sectionIndexEnabled ? header : nil

    let listSection = CPListSection.init(
      items: items, header: header, sectionIndexTitle: sectionIndexTitle)

    self._super = listSection
    return listSection
  }

  public func getFCPListTemplateItems() -> [FCPListTemplateItem] {
    return objcItems
  }
}
